File: ruby/packages/var-core/lib/oselvar/var/core/table_cells.rb
# frozen_string_literal: true

require 'oselvar/var/core/span'

module Oselvar
  module Var
    module Core
      # Parse a Markdown/Gherkin table row into trimmed cells and per-cell source
      # spans. Port of table-cells.ts. All offsets count UTF-16 code units.
      module TableCells
        module_function

        # Split a `| a | b |` row into [cells, cell_spans]. +line_start_offset+
        # is the UTF-16 offset of the row's first character within +source+.
        def parse_row_cells(line_text, line_start_offset, source)
          first_cp = line_text.index('|')
          last_cp = line_text.rindex('|')
          return [[], []] if first_cp.nil? || last_cp.nil? || last_cp <= first_cp

          # Pipe positions: convert code-point index to UTF-16 (matters when
          # astral chars precede the pipe). '|' is ASCII → 1 UTF-16 unit.
          first_u16 = Offsets.to_utf16_offset(line_text, first_cp)
          inner_start_u16 = first_u16 + 1

          inner = line_text[(first_cp + 1)...last_cp]

          cells = []
          cell_spans = []
          cursor = 0 # running UTF-16 position within inner

          # split(-1) keeps trailing empty segments, matching JS/Python split.
          inner.split('|', -1).each do |seg|
            trimmed = seg.strip
            leading = Offsets.utf16_len(seg) - Offsets.utf16_len(seg.lstrip)
            abs_start = line_start_offset + inner_start_u16 + cursor + leading
            cells << trimmed
            cell_spans << Offsets.span_from_offsets(source, abs_start, abs_start + Offsets.utf16_len(trimmed))
            cursor += Offsets.utf16_len(seg) + 1 # +1 for the '|' delimiter
          end

          [cells, cell_spans]
        end
      end
    end
  end
end

File: python/packages/var-core/src/varar_core/table_cells.py
"""
Parse a Markdown/Gherkin table row into trimmed cells and per-cell source spans.

Port of: typescript/packages/var-core/src/table-cells.ts

UTF-16 rule: all offsets (including line_start_offset and emitted Span offsets)
count UTF-16 code units, matching JavaScript's native string indexing so that
span values are identical between the TS and Python implementations.
"""

from varar_core.span import Span, span_from_offsets, utf16_len, to_utf16_offset


def parse_row_cells(
    line_text: str,
    line_start_offset: int,
    source: str,
) -> tuple[tuple[str, ...], tuple[Span, ...]]:
    """Split a ``| a | b |`` table row into trimmed cells and per-cell spans.

    Parameters
    ----------
    line_text:
        The raw text of the row (e.g. ``"| a | b |"``).
    line_start_offset:
        UTF-16 offset of the first character of *line_text* within *source*.
    source:
        The full document source (used by :func:`span_from_offsets` to compute
        line/column coordinates).

    Returns
    -------
    (cells, cell_spans)
        *cells* — tuple of trimmed cell strings.
        *cell_spans* — tuple of :class:`~var.span.Span` objects, one per cell,
        with ``start_offset``/``end_offset`` in UTF-16 units pointing at the
        trimmed cell text within *source*.
    """
    first_cp = line_text.find("|")
    last_cp = line_text.rfind("|")
    if first_cp < 0 or last_cp <= first_cp:
        return ((), ())

    # Convert code-point positions of the pipe characters to UTF-16 offsets
    # within line_text (matters when astral chars appear before the pipe).
    first_u16 = to_utf16_offset(line_text, first_cp)
    # inner_start is the UTF-16 offset within line_text of the char after the
    # opening '|'.  '|' is ASCII so it always takes exactly 1 UTF-16 unit.
    inner_start_u16 = first_u16 + 1

    # Extract the text between the first and last pipe using code-point slices
    # (Python slicing operates on code points, which is what we want for the
    # string values; we track UTF-16 widths separately for offset arithmetic).
    inner = line_text[first_cp + 1 : last_cp]

    cells: list[str] = []
    cell_spans: list[Span] = []
    cursor = 0  # running UTF-16 position within inner

    for seg in inner.split("|"):
        trimmed = seg.strip()
        # Number of leading-whitespace UTF-16 units (whitespace is always BMP,
        # but use utf16_len for correctness and symmetry with the TS port).
        leading = utf16_len(seg) - utf16_len(seg.lstrip())
        abs_start = line_start_offset + inner_start_u16 + cursor + leading
        cells.append(trimmed)
        cell_spans.append(
            span_from_offsets(source, abs_start, abs_start + utf16_len(trimmed))
        )
        cursor += utf16_len(seg) + 1  # +1 for the '|' delimiter itself

    return (tuple(cells), tuple(cell_spans))
